File: qualidade.js
'use strict';
// Same 24-hour price policy as integracoes/qualidade.py; checked in offline tests.
const MiraQuality = (() => {
    function usablePrice(product, now = Date.now()) {
        const e = product.priceCheck;
        if (!e || e.status !== 'verified' || e.method !== 'poly-card-v1' || e.itemId !== product.id || e.currency !== 'BRL') return false;
        if (typeof product.price !== 'number' || !Number.isFinite(product.price) || product.price <= 0 || product.available === false) return false;
        if (e.price !== product.price || e.oldPrice !== product.oldPrice) return false;
        if (typeof e.checkedAt !== 'string' || !/(Z|[+-]\d{2}:\d{2})$/.test(e.checkedAt)) return false;
        const age = now - Date.parse(e.checkedAt);
        return Number.isFinite(age) && age >= 0 && age <= 24 * 60 * 60 * 1000;
    }
    return {usablePrice};
})();
if (typeof module !== 'undefined') module.exports = MiraQuality;

File: integracoes/qualidade.py
"""Shared rules for using a price observation; historical data is not verified."""
from datetime import datetime, timezone
import math

MAX_PRICE_AGE_HOURS = 24


def usable_price(product, now=None):
    evidence = product.get('priceCheck') or {}
    if not isinstance(evidence, dict): return False
    if evidence.get('status') != 'verified' or evidence.get('method') != 'poly-card-v1': return False
    if evidence.get('itemId') != product.get('id') or evidence.get('currency') != 'BRL': return False
    price = product.get('price')
    if isinstance(price, bool) or not isinstance(price, (int,float)) or not math.isfinite(price) or price <= 0: return False
    if evidence.get('price') != price or evidence.get('oldPrice') != product.get('oldPrice'): return False
    if product.get('available') is False: return False
    try:
        checked = datetime.fromisoformat(evidence['checkedAt'].replace('Z','+00:00'))
        if checked.tzinfo is None: return False
        age = ((now or datetime.now(timezone.utc)) - checked).total_seconds()
    except (KeyError, TypeError, ValueError, AttributeError): return False
    return 0 <= age <= MAX_PRICE_AGE_HOURS * 3600
